### DISCUSSION TITLE
give CodeEdit node ability to set font for individual strings

### DISCUSSION THREAD

**Discussion by IllustratedMan-code** [godotengine/godot-proposals#4772] [GUI]
Currently, there is no way to set the font of a particular piece of text within a CodeEdit node. 

This would be very useful for code editors and the like, particularly for applications that imply style (markdown, LaTeX).  This might be best included as part of the CodeHighlighter node.

How easy to add would a feature like this be?

**Comment by Calinou** [comment]
Related to https://github.com/godotengine/godot-proposals/issues/1781 (which predates the TextEdit-CodeEdit split).

> How easy to add would a feature like this be?

It's difficult to add properly, as different fonts will have different metrics.

You can make it easier by only supporting fonts with the same metrics within a given CodeEdit (e.g. monospace or [uniwidth](https://www.recursive.design/)[^1] fonts only). This is usually good enough for code editors, but it won't suffice for those looking to create WYSIWYG editors, especially if they want to have different text sizes at the same time.

[^1]: An uniwidth font is a proportional (non-monospace) font whose weights and styles have the same metrics. These are especially suited to UI transitions as changing a font's weight or style will never cause an UI reflow.

**Reply by IllustratedMan-code** [reply]
I assume actually rendering the fonts is not the issue right? It's the positioning of the text that is the problem (calculating line height, cursor position, making sure text doesn't overlap, etc).

**Reply by Calinou** [reply]
Indeed. Individually rendering the fonts is fine, but managing spacing for different fonts can get complicated.

**Reply by IllustratedMan-code** [reply]
There does appear to be functionality for some of this already, just none of it is exposed.

the add_string method in text_paragraph.cpp [here](https://github.com/godotengine/godot/blob/af694245f51bb09a02812b0e2b5c700bad5a06ca/scene/resources/text_paragraph.cpp#L369) seems to take a font as an argument, implying that multiple fonts can exist simultaneously in a piece of text. It might not be that big of a job to expose some of the underlying functionality of the text server to the CodeEdit node (and perhaps TextEdit as well) to enable manipulation of the fonts from a script.

I'm pretty new to this codebase so I might be way off the mark in terms of how feasible this is.

**Reply by bruvzg** [reply]
Indeed, text server and TextLine / TextParagraph do support multiple fonts for any text. CodeEdit does not support it, most parts should work with it without any changes.

But code formatting is probably applied during draw, not when paragraph is updated, so it will require some changes to implement it.

**Reply by ghost** [reply]
An easier way to handle this would be to simply print the the inverted text after individualizing it (somehow), sort of like reverse templating, if you get what I mean.
